File: bindings/dotnet/src/Wcl/Serde/WclDeserializer.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using Wcl.Core;
using Wcl.Eval;

namespace Wcl.Serde
{
    public class SerdeError : Exception
    {
        public SerdeError(string message) : base(message) { }

        public static SerdeError TypeMismatch(string expected, string actual) =>
            new SerdeError($"type mismatch: expected {expected}, got {actual}");
        public static SerdeError MissingField(string name) =>
            new SerdeError($"missing required field: {name}");
    }

    public static class WclDeserializer
    {
        public static T FromValue<T>(WclValue value)
        {
            return (T)ConvertValue(value, typeof(T))!;
        }

        private static object? ConvertValue(WclValue value, Type targetType)
        {
            // Handle nullable
            var underlying = Nullable.GetUnderlyingType(targetType);
            if (underlying != null)
            {
                if (value.IsNull) return null;
                return ConvertValue(value, underlying);
            }

            // Null
            if (value.IsNull)
            {
                if (!targetType.IsValueType) return null;
                throw new SerdeError($"cannot assign null to {targetType.Name}");
            }

            // Primitives
            if (targetType == typeof(string))
            {
                return value.AsString();
            }
            if (targetType == typeof(long) || targetType == typeof(Int64)) return value.AsInt();
            if (targetType == typeof(int) || targetType == typeof(Int32)) return (int)value.AsInt();
            if (targetType == typeof(double)) return value.Kind == WclValueKind.Int ? (double)value.AsInt() : value.AsFloat();
            if (targetType == typeof(float)) return value.Kind == WclValueKind.Int ? (float)value.AsInt() : (float)value.AsFloat();
            if (targetType == typeof(bool)) return value.AsBool();

            // List<T>
            if (targetType.IsGenericType && targetType.GetGenericTypeDefinition() == typeof(List<>))
            {
                var itemType = targetType.GetGenericArguments()[0];
                var list = (IList)Activator.CreateInstance(targetType)!;
                foreach (var item in value.AsList())
                    list.Add(ConvertValue(item, itemType));
                return list;
            }

            // BlockRef -> Map with id auto-populated
            if (value.Kind == WclValueKind.BlockRef)
            {
                var br = value.AsBlockRef();
                var map = new OrderedMap<string, WclValue>();
                if (br.Id != null)
                    map["id"] = WclValue.NewString(br.Id);
                foreach (var kvp in br.Attributes)
                    map[kvp.Key] = kvp.Value;
                return ConvertValue(WclValue.NewMap(map), targetType);
            }

            // Set -> List coercion
            if (value.Kind == WclValueKind.Set && targetType.IsGenericType &&
                targetType.GetGenericTypeDefinition() == typeof(List<>))
            {
                return ConvertValue(WclValue.NewList(value.AsSet()), targetType);
            }

            // WclValue passthrough
            if (targetType == typeof(WclValue)) return value;

            // Dictionary<string, T>
            if (targetType.IsGenericType && targetType.GetGenericTypeDefinition() == typeof(Dictionary<,>))
            {
                var valType = targetType.GetGenericArguments()[1];
                var dict = (IDictionary)Activator.CreateInstance(targetType)!;
                if (value.Kind == WclValueKind.Map)
                {
                    foreach (var kvp in value.AsMap())
                        dict.Add(kvp.Key, ConvertValue(kvp.Value, valType));
                }
                return dict;
            }

            // POCO via reflection
            if (value.Kind == WclValueKind.Map)
            {
                var map = value.AsMap();
                var obj = Activator.CreateInstance(targetType)!;
                foreach (var prop in targetType.GetProperties(BindingFlags.Public | BindingFlags.Instance))
                {
                    if (!prop.CanWrite) continue;
                    var name = prop.Name;
                    var snakeName = ToSnakeCase(name);
                    if (map.TryGetValue(snakeName, out var val) || map.TryGetValue(name, out val))
                    {
                        prop.SetValue(obj, ConvertValue(val, prop.PropertyType));
                    }
                    else if (!IsOptionalType(prop.PropertyType))
                    {
                        throw SerdeError.MissingField(snakeName);
                    }
                }
                foreach (var field in targetType.GetFields(BindingFlags.Public | BindingFlags.Instance))
                {
                    var name = field.Name;
                    var snakeName = ToSnakeCase(name);
                    if (map.TryGetValue(snakeName, out var val) || map.TryGetValue(name, out val))
                    {
                        field.SetValue(obj, ConvertValue(val, field.FieldType));
                    }
                    else if (!IsOptionalType(field.FieldType))
                    {
                        throw SerdeError.MissingField(snakeName);
                    }
                }
                return obj;
            }

            throw new SerdeError($"cannot deserialize {value.TypeName} into {targetType.Name}");
        }

        private static bool IsOptionalType(Type type) =>
            !type.IsValueType || Nullable.GetUnderlyingType(type) != null;

        private static string ToSnakeCase(string name)
        {
            var sb = new System.Text.StringBuilder();
            for (int i = 0; i < name.Length; i++)
            {
                if (char.IsUpper(name[i]))
                {
                    if (i > 0) sb.Append('_');
                    sb.Append(char.ToLower(name[i]));
                }
                else
                {
                    sb.Append(name[i]);
                }
            }
            return sb.ToString();
        }
    }
}

File: bindings/jvm/src/main/java/io/github/wiltaylor/wcl/serde/WclDeserializer.java
package io.github.wiltaylor.wcl.serde;

import io.github.wiltaylor.wcl.eval.WclValue;
import io.github.wiltaylor.wcl.eval.WclValueKind;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.*;

public final class WclDeserializer {
    private WclDeserializer() {}

    @SuppressWarnings("unchecked")
    public static <T> T fromValue(WclValue value, Class<T> targetType) {
        return (T) convertValue(value, targetType, null);
    }

    @SuppressWarnings("unchecked")
    private static Object convertValue(WclValue value, Class<?> rawType, Type genericType) {
        if (value.isNull()) {
            if (rawType.isPrimitive()) {
                throw new SerdeException("cannot assign null to " + rawType.getName());
            }
            return null;
        }

        // Primitives and wrappers
        if (rawType == String.class) {
            if (value.getKind() == WclValueKind.DATE) return value.asDate();
            if (value.getKind() == WclValueKind.DURATION) return value.asDuration();
            return value.asString();
        }
        if (rawType == long.class || rawType == Long.class) {
            if (value.getKind() == WclValueKind.BIG_INT) return value.asBigInt();
            return value.asInt();
        }
        if (rawType == int.class || rawType == Integer.class) {
            if (value.getKind() == WclValueKind.BIG_INT) return (int) value.asBigInt();
            return (int) value.asInt();
        }
        if (rawType == double.class || rawType == Double.class) {
            return value.getKind() == WclValueKind.INT ? (double) value.asInt() : value.asFloat();
        }
        if (rawType == float.class || rawType == Float.class) {
            return value.getKind() == WclValueKind.INT ? (float) value.asInt() : (float) value.asFloat();
        }
        if (rawType == boolean.class || rawType == Boolean.class) return value.asBool();

        // WclValue passthrough
        if (rawType == WclValue.class) return value;

        // Object.class - auto-convert based on WclValue kind
        if (rawType == Object.class) {
            return switch (value.getKind()) {
                case STRING -> value.asString();
                case INT -> value.asInt();
                case FLOAT -> value.asFloat();
                case BOOL -> value.asBool();
                case NULL -> null;
                case LIST -> {
                    var items = value.asList();
                    var list = new ArrayList<>(items.size());
                    for (var item : items) list.add(convertValue(item, Object.class, null));
                    yield list;
                }
                case MAP -> {
                    var srcMap = value.asMap();
                    var result = new LinkedHashMap<String, Object>();
                    for (var entry : srcMap.entrySet())
                        result.put(entry.getKey(), convertValue(entry.getValue(), Object.class, null));
                    yield result;
                }
                case SET -> {
                    var items = value.asSet();
                    var list = new ArrayList<>(items.size());
                    for (var item : items) list.add(convertValue(item, Object.class, null));
                    yield list;
                }
                case BIG_INT -> value.asBigInt();
                case DATE -> value.asDate();
                case DURATION -> value.asDuration();
                case BLOCK_REF -> value.asBlockRef();
            };
        }

        // List<T>
        if (List.class.isAssignableFrom(rawType)) {
            Class<?> itemType = Object.class;
            if (genericType instanceof ParameterizedType pt) {
                var ta = pt.getActualTypeArguments()[0];
                if (ta instanceof Class<?> c) itemType = c;
            }
            var items = value.asList();
            var list = new ArrayList<>(items.size());
            for (var item : items) list.add(convertValue(item, itemType, null));
            return list;
        }

        // BlockRef -> Map with id auto-populated
        if (value.getKind() == WclValueKind.BLOCK_REF) {
            var br = value.asBlockRef();
            var map = new LinkedHashMap<String, WclValue>();
            if (br.getId() != null) map.put("id", WclValue.ofString(br.getId()));
            map.putAll(br.getAttributes());
            return convertValue(WclValue.ofMap(map), rawType, genericType);
        }

        // Set -> List coercion
        if (value.getKind() == WclValueKind.SET && List.class.isAssignableFrom(rawType)) {
            return convertValue(WclValue.ofList(value.asSet()), rawType, genericType);
        }

        // Map<String, T>
        if (Map.class.isAssignableFrom(rawType)) {
            Class<?> valType = Object.class;
            if (genericType instanceof ParameterizedType pt && pt.getActualTypeArguments().length >= 2) {
                var ta = pt.getActualTypeArguments()[1];
                if (ta instanceof Class<?> c) valType = c;
            }
            var srcMap = value.asMap();
            var result = new LinkedHashMap<>();
            for (var entry : srcMap.entrySet()) {
                result.put(entry.getKey(), convertValue(entry.getValue(), valType, null));
            }
            return result;
        }

        // POJO via reflection
        if (value.getKind() == WclValueKind.MAP) {
            var map = value.asMap();
            try {
                var obj = rawType.getDeclaredConstructor().newInstance();
                for (var field : rawType.getDeclaredFields()) {
                    if (java.lang.reflect.Modifier.isStatic(field.getModifiers())) continue;
                    field.setAccessible(true);
                    var snakeName = toSnakeCase(field.getName());
                    WclValue val = map.get(snakeName);
                    if (val == null) val = map.get(field.getName());
                    if (val != null) {
                        field.set(obj, convertValue(val, field.getType(), field.getGenericType()));
                    } else if (field.getType().isPrimitive()) {
                        throw new SerdeException("missing required field: " + snakeName);
                    }
                }
                return obj;
            } catch (SerdeException e) {
                throw e;
            } catch (Exception e) {
                throw new SerdeException("failed to deserialize into " + rawType.getName() + ": " + e.getMessage());
            }
        }

        throw new SerdeException("cannot deserialize " + value.typeName() + " into " + rawType.getName());
    }

    static String toSnakeCase(String name) {
        var sb = new StringBuilder();
        for (int i = 0; i < name.length(); i++) {
            var ch = name.charAt(i);
            if (Character.isUpperCase(ch)) {
                if (i > 0) sb.append('_');
                sb.append(Character.toLowerCase(ch));
            } else {
                sb.append(ch);
            }
        }
        return sb.toString();
    }
}
